lib/serverless/aws-lambda.js

```diff
@@ -82,8 +82,14 @@ class AwsLambda {
       // NOTE: This may be converted to holding onto a single ender function if only
       // one invocation is executing at a time.
       shim.wrap(process, 'emit', function wrapEmit(shim, emit) {
-        return function wrappedEmit(ev) {
-          if (ev === 'beforeExit') {
+        return function wrappedEmit(ev, error) {
+          // need to add error as uncaughtException to be used
+          // later to add to transaction errors
+          if (ev === 'unhandledRejection') {
+            uncaughtException = error
+          }
```

Comment:
When a promise is returned, we attempt to add a `.catch()` on it to grab the error and end the transaction. It then rethrows as we are trying not to tweak existing behavior too much. I'm wondering if that specific case would result in attempting to write this transactions data twice? We may have to remove the `.catch()` as I'd expect this to handle both cases.

Although, the test case you added I'd expect to have actually been caught via that case (see `lambdaInterceptPromise`). Would make sure we don't have the chance for accidentally grabbing the error /writing the payload more than once. I'm guessing not... and I'm also wondering when that intercept promise works if it not in the async-await case. Seems like it should detect a promise?

Reply:
I'll take a look at this...

>  I'm also wondering when that intercept promise works if it not in the async-await case. Seems like it should detect a promise?

So our instrumentation is checking the return value of the handler, so if that is a promise it does all that logic.  I can dig more and maybe get rid of some superfluous code now that we have a logic for `unhandledRejections`

Reply:
Yeah, its odd to me because in the case of async-await... the return value should be a promise. 🤷

Reply:
not if you don't return a promise.

Reply:
>  I'm wondering if that specific case would result in attempting to write this transactions data twice?

It does not.  Throwing in a catch does not cause an unhandled promise rejection.